### PR TITLE
Lowering for the Matmul Intrinsic

### PR DESCRIPTION
**Pull request by kiranchandramohan**

### CODE REVIEW

File: flang/lib/Lower/TransformationalRuntime.cpp [hunk 1]
```diff
@@ -18,6 +19,26 @@
...
+
+  func = Fortran::lower::getRuntimeFunc<mkRTKey(Matmul)>(loc, builder);
+  auto fTy = func.getType();
+  auto sourceFile = Fortran::lower::locationToFilename(builder, loc);
+  auto sourceLine =
+      Fortran::lower::locationToLineNo(builder, loc, fTy.getInput(4));
+  auto args =
+      Fortran::lower::createArguments(builder, loc, fTy, resultBox, matrixABox,
+                                      matrixBBox, sourceFile, sourceLine);
+  return builder.create<fir::CallOp>(loc, func, args).getResult(0);
```
Comment (jeanPerier) [inline flang/lib/Lower/TransformationalRuntime.cpp:39]: Shouldn't `genMatmul` return `void` ? `Matmul` runtime functions does not return anything.

Comment (kiranchandramohan) [reply flang/lib/Lower/TransformationalRuntime.cpp:39]: Yeah. :)

Comment (kiranchandramohan) [reply flang/lib/Lower/TransformationalRuntime.cpp]: Done.

File: flang/lib/Lower/IntrinsicCall.cpp [hunk 2]
```diff
@@ -2313,6 +2319,35 @@ IntrinsicLibrary::genCharacterCompare(mlir::Type type,
...
+  // Handle required matmul arguments
+  fir::BoxValue matrixTmpA = builder.createBox(loc, args[0]);
+  mlir::Value matrixA = fir::getBase(matrixTmpA);
+  fir::BoxValue matrixTmpB = builder.createBox(loc, args[1]);
+  mlir::Value matrixB = fir::getBase(matrixTmpB);
+
+  // Create mutable fir.box to be passed to the runtime for the result.
+  // TODO: The result can also be of rank 1 if one of the input matrices
+  // is of rank 1. Will this info be available at compile time or should
+  // code be generated to compute the rank?
```
[the remark this answers is not in the document]
Comment (kiranchandramohan) [reply flang/lib/Lower/IntrinsicCall.cpp]: Done.

**Review by jeanPerier — APPROVED** [review]
LGTM, thanks Kiran

**Comment by kiranchandramohan** [comment]
ping @mleair @schweitzpgi

**Review by mleair — APPROVED** [review]
LGTM

**Review by schweitzpgi — APPROVED** [review]
Thanks, Kiran.

**Pull request merged**